WordPress/wordpress-playground
# Add logging support to Playground

## What is this PR doing?

This PR adds support for collecting PHP logs and WordPress debug logs in Playground.
After each request logs are printed with `console.debug` in the browser.

Support for Node, WP-now, and VSCode will be added in separate PRs.
We also plan to add more Playground logs in the future to help us improve the stability of WordPress Playground.

## What problem is it solving?

When Playground crashes it's hard to know why it happened. By adding access to PHP logs and users will be able to debug their sites.
Playground developers will benefit from having Playground and PHP logs in a single place and a shared format.

## How is the problem addressed?

On `request.end` we collect logs from the current PHP request and send them back with the event.
These logs are caught by the Logger and printed.
Additionally, the logger catches some of the Playground errors using regular `window` events.

## Testing Instructions

- Checkout this branch
- Start Playground `npm run dev`
- [Open Playground in your browser](http://localhost:5400/website-server/?url=/wp-admin/post-new.php&networking=yes)
- Confirm that the debug log output is recorded in the browser console after each request
- Add `throw new Exception('This is a test exception');` after this file of https://github.com/WordPress/wordpress-playground/blob/c23dd927a92501fa94fd38d6eebc0d0ff39699eb/packages/playground/remote/src/lib/playground-mu-plugin/playground-includes/playground_logger.php#L9
- Refresh Playground and confirm that the PHP log is printed. This confirms that we can collect logs even if Playground fails to run
- Replace the Exception from /playground_logger.php with this code
```
add_action('init', function () {
    if (isset($_GET['post'])) {
        throw new Exception('This is a test exception');
    }
});
```
- Refresh Playground
- Confirm that it loads
- Go to a post edit page using wp-admin
- Confirm that the post edit page crashes and the debug log is printed
